File: camerax_android/android/src/main/kotlin/dev/hebei/camerax_android/common/ObserverImpl.kt
package dev.hebei.camerax_android.common

import androidx.camera.core.CameraState
import androidx.camera.core.ZoomState
import dev.hebei.camerax_android.CameraXImpl
import dev.hebei.camerax_android.PigeonApiCameraStateObserverApi
import dev.hebei.camerax_android.PigeonApiTorchStateObserverApi
import dev.hebei.camerax_android.PigeonApiZoomStateObserverApi
import dev.hebei.camerax_android.core.api
import dev.hebei.camerax_android.core.torchStateApi


class CameraStateObserverImpl(impl: CameraXImpl) : PigeonApiCameraStateObserverApi(impl) {
    override fun pigeon_defaultConstructor(): CameraStateObserver {
        return CameraStateObserver(this)
    }
}

class TorchStateObserverImpl(impl: CameraXImpl) : PigeonApiTorchStateObserverApi(impl) {
    override fun pigeon_defaultConstructor(): TorchStateObserver {
        return TorchStateObserver(this)
    }
}

class ZoomStateObserverImpl(impl: CameraXImpl) : PigeonApiZoomStateObserverApi(impl) {
    override fun pigeon_defaultConstructor(): ZoomStateObserver {
        return ZoomStateObserver(this)
    }
}

class CameraStateObserver(private val impl: CameraStateObserverImpl) {
    val obj = androidx.lifecycle.Observer<CameraState> { value ->
        impl.onChanged(this, value.api) {}
    }
}

class TorchStateObserver(private val impl: TorchStateObserverImpl) {
    val obj = androidx.lifecycle.Observer<Int> { value ->
        impl.onChanged(this, value.torchStateApi) {}
    }
}

class ZoomStateObserver(private val impl: ZoomStateObserverImpl) {
    internal val obj = androidx.lifecycle.Observer<ZoomState> { value ->
        impl.onChanged(this, value) {}
    }
}

File: camerax_ios/ios/Classes/common/ObserverImpl.swift
//
//  ObserverImpl.swift
//  camerax_ios
//
//  Created by 闫守旺 on 2025/5/15.
//

import Foundation

class TorchStateObserverImpl: PigeonApiDelegateTorchStateObserverApi {
    func pigeonDefaultConstructor(pigeonApi: PigeonApiTorchStateObserverApi) throws -> TorchStateObserver {
        return TorchStateObserver(pigeonApi)
    }
}

class ZoomStateObserverImpl: PigeonApiDelegateZoomStateObserverApi {
    func pigeonDefaultConstructor(pigeonApi: PigeonApiZoomStateObserverApi) throws -> ZoomStateObserver {
        return ZoomStateObserver(pigeonApi)
    }
}

class TorchStateObserver: Observer<TorchState> {
    typealias T = TorchState
    
    let api: PigeonApiTorchStateObserverApi
    
    init(_ api: PigeonApiTorchStateObserverApi) {
        self.api = api
    }
    
    func onChanged(_ value: TorchState) {
        api.onChanged(pigeonInstance: self, value: value.api) { _ in }
    }
}

class ZoomStateObserver: Observer<ZoomState> {
    typealias T = ZoomState
    
    let api: PigeonApiZoomStateObserverApi
    
    init(_ api: PigeonApiZoomStateObserverApi) {
        self.api = api
    }
    
    func onChanged(_ value: ZoomState) {
        api.onChanged(pigeonInstance: self, value: value) { _ in }
    }
}
